src/mjolnir/graphfilter.cc

```diff
@@ -219,7 +393,359 @@ void FilterTiles(GraphReader& reader,
+      uint32_t edge_index = tilebuilder.directededges().size();
+
+      // Iterate through directed edges outbound from this node
+      std::vector<uint64_t> wayid;
+      std::vector<GraphId> endnode;
+      const NodeInfo* nodeinfo = tile->node(nodeid);
+
+      // Nodes marked with mode_change = true are tossed.
+      if (nodeinfo->mode_change()) {
+        continue;
+      }
+
+      GraphId edgeid(nodeid.tileid(), nodeid.level(), nodeinfo->edge_index());
+
+      for (uint32_t j = 0; j < nodeinfo->edge_count(); ++j, ++edgeid) {
+        // Check if the directed edge should be included
+        const DirectedEdge* directededge = tile->directededge(edgeid);
+
+        // Copy the directed edge information
+        DirectedEdge newedge = *directededge;
+
+        // Set opposing edge indexes to 0 (gets set in graph validator).
+        newedge.set_opp_index(0);
+
+        // Get signs from the base directed edge
+        if (directededge->sign()) {
+          std::vector<SignInfo> signs = tile->GetSigns(edgeid.id());
+          if (signs.size() == 0) {
+            LOG_ERROR("Base edge should have signs, but none found");
+          }
+          tilebuilder.AddSigns(tilebuilder.directededges().size(), signs);
+        }
+
+        // Get turn lanes from the base directed edge
+        if (directededge->turnlanes()) {
+          uint32_t offset = tile->turnlanes_offset(edgeid.id());
+          tilebuilder.AddTurnLanes(tilebuilder.directededges().size(), tile->GetName(offset));
+        }
+
+        // Get access restrictions from the base directed edge. Add these to
+        // the list of access restrictions in the new tile. Update the
+        // edge index in the restriction to be the current directed edge Id
+        if (directededge->access_restriction()) {
+          auto restrictions = tile->GetAccessRestrictions(edgeid.id(), kAllAccess);
+          for (const auto& res : restrictions) {
+            tilebuilder.AddAccessRestriction(AccessRestriction(tilebuilder.directededges().size(),
+                                                               res.type(), res.modes(), res.value()));
+          }
+        }
+
+        // Copy lane connectivity
+        if (directededge->laneconnectivity()) {
+          auto laneconnectivity = tile->GetLaneConnectivity(edgeid.id());
+          if (laneconnectivity.size() == 0) {
+            LOG_ERROR("Base edge should have lane connectivity, but none found");
+          }
+          for (auto& lc : laneconnectivity) {
+            lc.set_to(tilebuilder.directededges().size());
+          }
+          tilebuilder.AddLaneConnectivity(laneconnectivity);
+        }
+
+        // Names can be different in the forward and backward direction
+        diff_names = tilebuilder.OpposingEdgeInfoDiffers(tile, directededge);
+
+        bool added;
+        auto edgeinfo = tile->edgeinfo(directededge);
+        std::string encoded_shape = edgeinfo.encoded_shape();
```

Comment:
we can probably avoid this copy if we're just overwriting it with the new encoded shape just below?

Reply:
We need the starting shape to build the final shape as we aggregate the edge.  So yes I need to have a copy of this as if we aggregate we update it.